Fixes to enable Windows CNI

madhanrm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
This PR has fixed which enables Kubelet to use Windows CNI plugin.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
#49646 
**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

madhanrm committed:
Sandbox Support for Windows + CNI

Following are part of this commit
+++++++++++++++++++++++++++++++++

* Windows CNI Support
	(1) Support to use --network-plugin=cni
	(2) Handled platform requirement of calling CNI ADD for all the containers.
	(2.1) For POD Infra container, netNs has to be empty
	(2.2) For all other containers, sharing the network namespace of POD container,
	      should pass netNS name as "container:<Pod Infra Container Id>", same as the
              NetworkMode of the current container
	(2.3) The Windows CNI plugin has to handle this to call into Platform.
              Sample Windows CNI Plugin code to be shared soon.
* Sandbox support for Windows
	(1) Sandbox support for Windows. Works only with Docker runtime.
	(2) Retained CONTAINER_NETWORK as a backward compatibilty flag,
	    to not break existing deployments using it.
	(3) Works only with CNI plugin enabled.

(*) Changes to reinvoke CNI ADD for every new container created. This is hooked up with PodStatus,
    but would be ideal to move it outside of this, once we have CNI GET support

madhanrm commented:
Do not merge yet.
There is still e2e test pending

freehan commented:
/ok-to-test

dineshgovindasamy commented:
Tagging @jdumars  for FYI

squeed commented on pkg/kubelet/network/cni/cni_windows.go:
```diff
@@ -0,0 +1,64 @@
+	return nil
+}
+
+// GetPodNetworkStatus : Assuming addToNetwork is idempotent, we can call this API as many times as required to get the IPAddress
```
`addToNetwork` is definitely not idempotent. The CNI spec doesn't require it and no plugins that I'm aware of implement ADD in this manner. The comment from before is... optimistic.

We're currently working on a spec change for a GET method within CNI.

squeed replied:
But it's a plugin system. You cannot guarantee that users will use your plugin.

alinbalutoiu replied:
I agree with squeed here.

madhanrm replied:
Well, let me put it this way. In windows, There is no other way or tools like nsenter available to get the IPAddress of the endpoint attached to the container in a namespace, other than calling the CNI Plugin again. The concepts are different for Windows. It would be better for the Plugin which allocated an endpoint, to get back the IPAddress of the endpoint it manages. It may not be feasible to find some alternative ways to get the endpoint ipaddress, like trying to use some other inbuilt tools. There are other complications there. 
Hence it is made idempotent here. This is a requirement for Windows CNI Plugins.
Anyone writing a windows CNI plugin, has to know this platform requirement, which would be published separately, somewhere. 

There is another requirement of windows CNI plugin, which is different from the Linux CNI plugin, like the CNI plugin has to be invoked with ADD for second or subsequent containers created on a POD, with netNsName as the "container:<Infra POD ContainerID>". This is again a platform requirement to replicate some of the critical data to the new container. Comment for this is added in the code itself.

Plugins are written this way to that platform specific requirements could be offloaded to the Plugin rather than putting them in the Kubernetes code, following the CNI spec.

squeed replied:
If the plugin can "enter the namespace and list the IPs," then why can't Kubernetes?

madhanrm replied:
The Plugin doesn't enter the namespace and get the IP directly. It calls into Host Networking Service, to get the data. Kubernetes can also do that directly, but then we would destroy the very idea of abstracting that logic out behind the plugin and not bringing in a lot of platform specific APIs.

The main reason behind this approach is that, current containers are launched with process isolation. It could also be launched with hyper-v isolation (not supported in kubernetes but supported by docker and platform). To keep the implementation same across these modes, it is better to call the centralized to CNI plugin to get back the data, the current API would seamlessly support these.

dcbw replied:
@madhanrm I agree with @squeed.  The CNI specification does not declare ADD idempotent.  That has nothing to do with the platform at all, that's just how it is.  Making Windows act differently is simply confusing.  So can we discuss this a bit more so that we all understand the restrictions?

Saying "This is a requirement for Windows CNI Plugins." is a solution to a problem, but there may be other solutions that do not require making ADD idempotent for Windows CNI.

IIRC Windows has no shared "infra" (eg net namespace) container.  Instead, *every* container gets network setup called on it.  What I think we're arguing about here, and correct me if I'm wrong, is that in the current PR here we'll get an ADD for the same container/namespace/netns tuple multiple times, which would in fact require idempotency that CNI does not guarantee.

Instead of that, could we figure out some other way to pass the data that your Windows plugin requires to tie the containers together?

One other thing I'm not clear on.  What actual CNI plugin are you working on?  And why couldn't some other 3rd party Windows plugin have a completely different mechanism for tying containers together, that doesn't require ADD to be idempotent?

dnardo replied:
@dcbw Just for my own understanding, what is the argument against making ADD idempotent in the spec?

madhanrm replied:
@dcbw 
The current Idempotent nature of ADD is only required because any subsequent ADD is used as a GET (PodStatus) to get the IP. Once we have GET implemented, that will not be a criteria for ADD to be idempotent. (But IMO, it doesn't make sense for ADD not to be idempotent, which is a separate discussion altogether)

Also, your understanding of "every container gets network setup" is getting changed by this PR. This PR is also to start using shared infra container, for which CNI is a requirement.

You should also understand the difference between the CNI plugin in Linux and in Windows. You really do not have to write individual CNI plugins on Windows unlike Linux. Instead CNI plugin is like a proxy to program the Host Networking Service (HNS) on Windows, which takes care of everything. The information is simply constructed and passed on to HNS. So a single CNI plugin would work for most of the network modes that can work independently on a single host. Other may develop similar plugins, but they have to go thro HNS, anything other than that would not be supported in Windows.
The idempotency is required here, because the CNI plugin doenst have to manage anything and simply pass it on to HNS. Else, the CNI would also have to manage redundant data as HNS. This would simplify the requirement of the Windows CNI plugins.

madhanrm replied:
Are there any open items here? Can we not relax the idempotent nature of ADD for Windows here?
Am not seeing any progress in this PR.

squeed commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -319,6 +263,10 @@ func (plugin *cniNetworkPlugin) buildCNIRuntimeConf(podName string, podNs string
 	glog.V(4).Infof("Using netns path %v", podNs)
 
+	if podNetnsPath == "" {
+		podNetnsPath = "none"
```
"none" is theoretically a valid netns path on Linux - can you just use the empty string?

madhanrm replied:
Sure, I can remove this change.

jdumars commented:
/sig windows

jdumars commented:
/sig node

jdumars commented:
/sig network

jdumars commented:
@caseydavenport could you or a delegate from SIG Network take a look at this implementation + #51064 

Thanks!

freehan commented:
Thanks for the PR! 

This come up late in the 1.8 cycle and I am not comfortable to put it in. There is no test coverage for windows cni plugin and the test coverage for linux cni plugin is not great either. I would recommend we get it in during early 1.9 cycle and improve test coverage. I would also recommend you to join the sig-network meeting to get involved in feature planning. That would help the feature that you are working on to gain more visibility in the k8s community. 

Same applies to #51064

JMesser81 commented:
@freehan - if you look at the code, these are actually minor changes and we've isolated by separating into Windows and Linux helper function. AFAIK, there is no CNI plugin support on Windows today so this is net-new. We are very close to getting tests to pass. 

Can we discuss in sig-network meet-up? This and #51064 ?

madhanrm commented on pkg/kubelet/network/cni/cni_windows.go:
```diff
@@ -0,0 +1,64 @@
+	return nil
+}
+
+// GetPodNetworkStatus : Assuming addToNetwork is idempotent, we can call this API as many times as required to get the IPAddress
```
The plugin that I have is implemented in that way, assuming no one would require adding 2 endpoints of the same type. This also simplifies the implementation of the windows cni plugin, which doesn't have to keep a repo of these data and simply pass these down to the platform based on those assumptions.

madhanrm replied:
Also, as soon as GET method is in CNI, we can switch GetPodNetworkStatus to use the new API.

madhanrm commented:
Currently working on getting these test passed. For Windows specific changes, these have been extensively tested locally. We need to work on getting a testing suite for windows though, since am not sure if one is already available.

madhanrm commented:
go test -v k8s.io/kubernetes/pkg/api -run TestRoundTripTypes$ => Passes for me. Am not sure why it is failing here.

        roundtrip.go:216:       round tripping to storage.k8s.io/v1, Kind=StorageClassList v1.StorageClassList
        roundtrip.go:137: finished group "storage.k8s.io"
PASS
ok      k8s.io/kubernetes/pkg/api       3.723s

madhanrm commented:
/test pull-kubernetes-e2e-kops-aws

madhanrm commented:
/test pull-kubernetes-unit

aserdean commented:
(This is not an actual technical review, but it would make it easier
reviewing the PR)

Please add commit messages to each individual commit, it makes it easier for
other people understand what and how you try to achieve your goal.

Please try to make each commit a unique logical change.
It should also reflect what you indicated in the commit title and message.
(eg : https://github.com/kubernetes/kubernetes/pull/51063/commits/7638ae118229dade86fbfa2cc4a3dc46e0c65723
besides the Powershell scripts you change some code from unrelated files)

Please write the commit titles in imperative
(eg: "Windows CNI Support" could be "Add CNI support for Windows").

Doing all of the above makes it easier to understand the evolution of the code.
The "why" is more important in the commit message than the "what" - the "what"
is already stated in the code.

Linus wrote some nice guidelines regarding titles and commit messages here:
https://github.com/torvalds/subsurface/blob/a48494d2fbed58c751e9b7e8fbff88582f9b2d02/README#L91

madhanrm commented:
/test pull-kubernetes-e2e-kops-aws

alinbalutoiu reviewed (commented):
I aggree to @aserdean, the code needs to be cleaned up a bit before the review. It's confusing the way it is splitted now in commits.

jdumars commented:
/cc @thockin

madhanrm commented:
Updating Commit messages and adding more comments in the code.

madhanrm commented on pkg/kubelet/dockershim/helpers_windows.go:
```diff
@@ -74,11 +74,20 @@ func (ds *dockerService) determinePodIPBySandboxID(sandboxID string) string {
+
+		// Using the CONTAINER_NETWORK flag to determine to choose old impl or current impl.
+		// CONTAINER_NETWORK is only kept for backward compatibility, since old OS version doens;t support
+		// Sandbox concept
```
Done.

madhanrm commented:
/test pull-kubernetes-federation-e2e-gce

madhanrm commented:
/test pull-kubernetes-federation-e2e-gce

thockin commented:
I am ill equipped to review this.  Please assign to me if this gets LGTM and needs approval.

dcbw commented on pkg/kubelet/dockershim/helpers_unsupported.go:
```diff
@@ -47,3 +47,11 @@ func (ds *dockerService) determinePodIPBySandboxID(uid string) string {
 	return ""
 }
+
+func getNetworkNamespace(c *dockertypes.ContainerJSON) (string, error) {
```
This function is actually pretty Linux specific.  Might as well just return an error for _unsupported.

dineshgovindasamy replied:
This has been fixed in our latest commit

dcbw commented on pkg/kubelet/dockershim/helpers_windows.go:
```diff
@@ -71,14 +73,35 @@ func (ds *dockerService) determinePodIPBySandboxID(sandboxID string) string {
-			return containerIP
+
+		// On Windows, every container that is created in a Sandbox, needs to invoke CNI plugin again for adding the Network,
+		// with the shared container name as NetNS info,
```
This comment doesn't square with getNetworkNamespace() below, which appears to return the network mode (not a shared container name) as the NetNS info.  Or am I reading it wrong?

madhanrm replied:
Yes, it returns the networkmode, which is container:<shared containername>

dcbw commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -153,43 +152,6 @@ func vendorCNIDir(prefix, pluginType string) string {
-	return loNetwork
-}
-
-func (plugin *cniNetworkPlugin) Init(host network.Host, hairpinMode kubeletconfig.HairpinMode, nonMasqueradeCIDR string, mtu int) error {
```
I'd actually rather have the Init() function left in cni.go, and then call a platform-specific "platformInit()" function or something like that and return an error if oyu get one back from it.  At least half of the existing Init() is platform-independent.  The only bits that aren't are checking for nsenter (which I actually want to kill anyway, once we have CNI GET support).

madhanrm replied:
Sure. Makes sense.

feiskyer commented on pkg/kubelet/dockershim/helpers_windows.go:
```diff
@@ -71,14 +73,35 @@ func (ds *dockerService) determinePodIPBySandboxID(sandboxID string) string {
+		// Using the CONTAINER_NETWORK flag to determine to choose old impl or current impl.
+		// CONTAINER_NETWORK is only kept for backward compatibility, since old OS version doens;t support
+		// Sandbox concept
+		if networkMode := os.Getenv("CONTAINER_NETWORK"); networkMode == "" {
```
weird of using a environment variable to determine kubelet's behavior, which is actually depending on the OS version. Is it possible of getting os version instead?

madhanrm replied:
Yes, will fix it separately. In my TODO list.

yujuhong replied:
Can you at least document what version relies on this env variable?

madhanrm commented:
/test pull-kubernetes-bazel-build
/test pull-kubernetes-bazel-test

rphillips commented on pkg/kubelet/network/cni/cni_linux.go:
```diff
@@ -0,0 +1,84 @@
+
+func (plugin *cniNetworkPlugin) Init(host network.Host, hairpinMode kubeletconfig.HairpinMode, nonMasqueradeCIDR string, mtu int) error {
+	var err error
+	plugin.nsenterPath, err = plugin.execer.LookPath("nsenter")
```
It would be more idiomatic to assign to a temporary variable here, then assign the temporary variable to plugin.nsenterPath after the if check

madhanrm replied:
Not related to my changes. This was just a code move to a different file.

madhanrm commented:
/retest

madhanrm commented:
/rerun

alinbalutoiu commented on cluster/windows/cni/kubeproxystart.ps1:
```diff
@@ -1,3 +1,18 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
```
Does this really require a separate commit to only add the licenses?

madhanrm replied:
Squeezed this commit. Was just following the guidance to not to squeeze until the final merge, to ease the review process.

madhanrm replied:
Taking care of the other comments - removing the ps1 files & updating the commit message.

alinbalutoiu reviewed (commented):
You should separate the powershell code from the go code and put it in another commit.

I think that the following structure would be the best:
- one commit with the golang changes (the changes from the commit with fixes bazel build could be moved here as well);
- another one with the powershell scripts and license (single commit, not separate commits);

Ideally the powershell code should be moved in a separate PR since they are not essential to getting CNI working on Windows. That would ease the review process of this PR in my opinion.

Please also insert a new line between the commit title and the commit message. All the lines until the new line are considered to be a part of the title. You could also use a more suggestive commit title for that one.

madhanrm commented:
Is there a way to fix the cross build for Darwin. I have added a build tag for cni_linux.go to // +build !windows. I expect this to work with all OS except windows. Why doenst it work for Darwin? Any clues?

madhanrm commented:
/rerun

madhanrm commented:
/retest

madhanrm commented:
/retest

madhanrm commented:
/retest

madhanrm commented:
Lot of review comments went missing after a rebase. How do I get back all of them?

madhanrm commented:
/retest

madhanrm commented:
/retest

dineshgovindasamy commented:
@k8s-ci-robot : 
Signed up with CCNF, can you please recheck?

dineshgovindasamy commented:
@alinbalutoiu In my latest commit i have removed the powershell code from the GO code.

alinbalutoiu commented:
@dineshgovindasamy can you please squash all the commits into a single commit? There's no point in adding the code in one commit and removing it in another commit.

dineshgovindasamy commented:
/retest

dineshgovindasamy commented:
/retest

dineshgovindasamy commented:
test /all

dineshgovindasamy commented:
/retest

dineshgovindasamy commented:
/retest

jdumars commented:
/lgtm

dineshgovindasamy commented:
/assign @dchen1107

Could you please review and approve this PR? We have fixed most of the comments. Some of them we would like to handle them as separate PR.

madhanrm commented:
/assign @thockin

madhanrm commented:
/retest

thockin commented:
@dcbw - are you LGTM on this one?

dineshgovindasamy commented:
@dcbw Could you please take a look?

dineshgovindasamy commented:
@thockin I haven't received any response from @dcbw . Many people are waiting on this windows CNI PR. Is it possible to assign someone else please?

squeed commented:
Dan (dcbw) is basically OOO for the next month.

thockin commented:
@freehan - do you have time to look at this?

dineshgovindasamy commented:
@freehan Could you please review this PR?

alinbalutoiu commented:
@dineshgovindasamy can you please fix the merge conflicts? Thanks!

dineshgovindasamy commented:
Rebased to the master

dineshgovindasamy commented:
/retest

dineshgovindasamy commented:
/retest

cmluciano commented:
@kubernetes/sig-network-pr-reviews

dineshgovindasamy commented:
@thockin  @freehan Could you please review this PR?

freehan commented:
Will take a look today. Thanks!

freehan commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -238,11 +213,12 @@ func (plugin *cniNetworkPlugin) SetUpPod(namespace string, name string, id kubec
-	if err != nil {
-		glog.Errorf("Error while adding to cni lo network: %s", err)
-		return err
+	if plugin.loNetwork != nil {
```
nit: add a comment to say linux comes with loNetwork and windows doesn't.

madhanrm replied:
Taken care of

freehan commented:
Just a nit.
/approve

At some point, we might just want to separate cni windows and linux. But it still seems maintainable for now.

yujuhong commented on pkg/kubelet/dockershim/helpers_windows.go:
```diff
@@ -71,14 +73,35 @@ func (ds *dockerService) determinePodIPBySandboxID(sandboxID string) string {
 	return ""
 }
 
+func getNetworkNamespace(c *dockertypes.ContainerJSON) (string, error) {
```
Please add a comment explaining what you're returning here and why. I assume this is not the "namesapce"

madhanrm replied:
done

yujuhong commented on pkg/kubelet/dockershim/helpers_windows.go:
```diff
@@ -71,14 +73,35 @@ func (ds *dockerService) determinePodIPBySandboxID(sandboxID string) string {
+		// with the shared container name as NetNS info,
+		// which is passed down the platform to replicate some necessary changes on to the new container
+		//
+		// This place is chosen as a hack for now, until we have a better place to do this,
```
Please file an issue for this and add the link here.

yujuhong commented on pkg/kubelet/dockershim/helpers_windows.go:
```diff
@@ -47,6 +47,8 @@ func (ds *dockerService) updateCreateConfig(
 	if networkMode := os.Getenv("CONTAINER_NETWORK"); networkMode != "" {
 		createConfig.HostConfig.NetworkMode = dockercontainer.NetworkMode(networkMode)
+	} else {
+		modifyHostNetworkOptionForContainer(false, podSandboxID, createConfig.HostConfig)
```
This function is in `security_context.go`. Calling this in directly seems weird. I understand that this part of code for windows and linux are very different, maybe refactor/clean this up in the future.

yujuhong commented on pkg/kubelet/network/cni/cni_windows.go:
```diff
@@ -0,0 +1,59 @@
+		glog.Errorf("Error while cni parsing result: %s", err)
+		return nil, err
+	}
+	// Parse the result and get the IPAddress
```
nit: I don't see much parsing here, so perhaps you can just remove the comment.

yujuhong replied:
Or maybe the comment was meant for line 51-55?

yujuhong commented on pkg/kubelet/network/cni/cni_windows.go:
```diff
@@ -0,0 +1,59 @@
+
+	result, err := plugin.addToNetwork(plugin.getDefaultNetwork(), name, namespace, id, netnsPath)
+
+	glog.V(5).Infof("GetPodNetworkStatus result %+v", result)
```
Why printing the result before checking the error?

madhanrm replied:
Would like to know the result either way. Error could contain some meaningful result to investigate issues.

yujuhong commented on pkg/kubelet/dockershim/helpers_windows.go:
```diff
@@ -71,14 +73,35 @@ func (ds *dockerService) determinePodIPBySandboxID(sandboxID string) string {
+		//
+		// This place is chosen as a hack for now, until we have a better place to do this,
+		// like immediately after ContainerCreation,
+		// Also this is very windows specific for now
```
remove `Also this is very windows specific for now`
I think it's pretty clear from the context that this is windows specific.

yujuhong commented:
Mostly nits since @freehan already lgtm'd it. 

Please ping me for approval after the comments are addressed.

madhanrm commented on pkg/kubelet/dockershim/helpers_unsupported.go:
```diff
@@ -19,6 +19,7 @@ limitations under the License.
 package dockershim
 
 import (
+	"fmt"
```
Done

madhanrm commented:
@yujuhong I have addressed all the comments. But while testing I found an issue in the latest mainline code, where os.Rename is failing with file already in use.
I see that util/store/filestore.go is newly added and  func writeFile(fs utilfs.Filesystem, path string, data []byte) (retErr error)  is calling fs.Rename at the end of the method, without closing the file (tmpFile.Close()). File closing is done in defer.

Can you check?

madhanrm commented on pkg/kubelet/network/cni/cni_windows.go:
```diff
@@ -0,0 +1,59 @@
+package cni
+
+import (
+	"fmt"
```
done

yujuhong commented:
> Can you check?

Oops...I sent a fix #55034. Sorry about that.

madhanrm commented:
/test pull-kubernetes-node-e2e

madhanrm commented:
@yujuhong  Validated your fix. Thanks for the quick response.
Can you approve this PR?

yujuhong commented:
> @yujuhong I have addressed all the comments. But while testing I found an issue in the latest mainline code, where os.Rename is failing with file already in use.
> I see that util/store/filestore.go is newly added and func writeFile(fs utilfs.Filesystem, path string, data []byte) (retErr error) is calling fs.Rename at the end of the method, without closing the file (tmpFile.Close()). File closing is done in defer.

@madhanrm were you testing on Windows?

yujuhong commented:
/approve 
based on @freehan's LGTM

madhanrm commented:
@freehan @yujuhong can one of you add a lgtm label.

freehan commented:
/lgtm
